internal/annotators/http/pki_test.go

```diff
@@ -0,0 +1,182 @@
+/*******************************************************************************
+ * Copyright 2022 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package http
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/project-alvarium/alvarium-sdk-go/internal/signprovider/ed25519"
+	"github.com/project-alvarium/alvarium-sdk-go/pkg/contracts"
+
+	"github.com/project-alvarium/alvarium-sdk-go/internal/annotators"
+	"github.com/project-alvarium/alvarium-sdk-go/pkg/config"
+	"github.com/project-alvarium/alvarium-sdk-go/test"
+)
+
+func TestHttpPkiAnnotator_Do(t *testing.T) {
+	b, err := ioutil.ReadFile("./test/config.json")
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	var cfg config.SdkInfo
+	err = json.Unmarshal(b, &cfg)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	badKeyType := cfg
+	badKeyType.Signature.PublicKey.Type = "invalid"
+
+	keyNotFound := cfg
+	keyNotFound.Signature.PublicKey.Path = "/dev/null/private.key"
+
+	req, data, err := buildRequest(cfg.Signature)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	ctx := context.WithValue(req.Context(), testRequest, req)
+
+	//Building a new request with an empty signature
+	req2 := req.Clone(req.Context())
+	req2.Header.Set("Signature", "")
+	ctx2 := context.WithValue(req2.Context(), testRequest, req2)
+
+	tests := []struct {
+		name        string
+		ctx         context.Context
+		cfg         config.SdkInfo
+		expectError bool
+	}{
+		{"pki annotation OK", ctx, cfg, false},
+		{"pki bad key type", ctx, badKeyType, true},
+		{"pki key not found", ctx, keyNotFound, true},
+		{"pki empty signature", ctx2, cfg, false},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pki := NewHttpPkiAnnotator(tt.cfg)
+			anno, err := pki.Do(tt.ctx, data)
+			test.CheckError(err, tt.expectError, tt.name, t)
+			if err == nil {
+				result, err := annotators.VerifySignature(tt.cfg.Signature.PublicKey, anno)
```

Comment:
@husseinfakharany Not sure this test is correct. This `VerifySignature` in base.go validates the signature on the Annotation itself. It doesn't refer to the signature in the Http Header. The latter is validated (or should be) when you call `pki.Do` above.

From `http/pki.go`
```
	var sig signable
	sig.Seed = parsed
	sig.Signature = req.(*http.Request).Header.Get("Signature")

	ok, err := sig.verifySignature(a.sign.PublicKey)
```
In this case, `verifySignature` is on the signable type -- and this ties to my other comment about extending the signable type so you can obtain the algorithm and keyid via the parser.

Reply:
@tsconn23 I've added a couple more lines to verify the isSatisfied boolean, and added more test cases.
To do so, I've changed the testing structure a bit, a `buildContext` function is created which takes a test name and the the request built in `line 52` as a params. This function returns a specific context for every single test case using a clone of the HTTP request to be used by the `pki.Do`. Regarding the "bad key type" and "key not found" tests, I've used a regular expression to replace the value of the `alg` and the `keyid` parameters already present in the `Signature-Input` header with "invalid".